### PR TITLE
Check for non-utf8 characters. Fix #1197

### LINKED ISSUE #1197: UTF-8 encoding errors
When non UTF-8 readable characters are inserted into the prep or sample template, the SQL insertion of the data goes haywire and returns an ugly SQL error.

### PR DESCRIPTION
**Pull request by squirrelo**
Does what it says on the tin. Also raises meaningful error with location of cells that have non-utf8 characters.

### CODE REVIEW

File: qiita_db/metadata_template/util.py [hunk 1]
```diff
@@ -215,21 +216,33 @@ def load_template_to_dataframe(fn, strip_whitespace=True):
...
+        # Find row number and col number for utf-8 encoding errors
+        errors = []
+        for row, line in enumerate(holdfile):
+            for col, cell in enumerate(line.split('\t')):
+                try:
+                    cell.encode('utf-8')
+                except UnicodeError:
+                    errors.append('Non-unicode value in cell at '
+                                  'row %d col %d' % (row+1, col+1))
+        raise ValueError(', '.join(errors))
```
Comment (josenavas) [inline qiita_db/metadata_template/util.py]: Can you rise one of the QiitaDB errors?

Comment (josenavas) [inline qiita_db/metadata_template/util.py]: And update the documentation of the method to reflect this change?

**Comment by josenavas** [comment]
Just a small comment @squirrelo

File: qiita_db/metadata_template/util.py (hunk 1 above)
Comment (squirrelo) [inline qiita_db/metadata_template/util.py]: None of the QiitaDB errors make much sense here though. Which would you suggest?

**Comment by antgonza** [comment]
Code looks good, could you add a test?

**Comment by squirrelo** [comment]
Oh, yup, good call.

File: qiita_db/metadata_template/util.py (hunk 1 above)
Comment (josenavas) [inline qiita_db/metadata_template/util.py]: Just use QiitaDBError, at least we know that the errors is generated by us.

File: qiita_db/metadata_template/util.py [hunk 2]
```diff
@@ -215,21 +215,33 @@ def load_template_to_dataframe(fn, strip_whitespace=True):
...
+                                   # collection_timestamp is not added here
+                                   'host_subject_id': str,
+                                   'description': str,
+                                   # common prep template information
+                                   'center_name': str,
+                                   'center_projct_name': str})
+    except UnicodeDecodeError:
+        # Find row number and col number for utf-8 encoding errors
+        errors = []
+        for row, line in enumerate(holdfile):
```
Comment (wasade) [inline qiita_db/metadata_template/util.py]: suggest using `enumerate(holdfile, 1)` given the use of the index below. Same on the nested forloop

File: qiita_db/metadata_template/util.py [hunk 3]
```diff
@@ -215,21 +215,33 @@ def load_template_to_dataframe(fn, strip_whitespace=True):
...
+            for col, cell in enumerate(line.split('\t')):
+                try:
+                    cell.encode('utf-8')
+                except UnicodeError:
+                    errors.append('Non-unicode value in cell at '
```
Comment (wasade) [inline qiita_db/metadata_template/util.py]: This text is redundant and doesn't add anything. The message could just be stated once when raising the specific exception below

File: qiita_db/metadata_template/util.py [hunk 4]
```diff
@@ -215,21 +215,33 @@ def load_template_to_dataframe(fn, strip_whitespace=True):
...
+                                  'row %d col %d' % (row+1, col+1))
```
Comment (wasade) [inline qiita_db/metadata_template/util.py]: the col number in particular will be difficult for someone using excel, what about presenting the user `(sample_name, header)`?

File: qiita_db/metadata_template/util.py (hunk 2 above)
Comment (squirrelo) [inline qiita_db/metadata_template/util.py]: Nice, didn't know about that. thanks.

File: qiita_db/metadata_template/util.py (hunk 3 above)
Comment (squirrelo) [inline qiita_db/metadata_template/util.py]: Done

File: qiita_db/metadata_template/util.py (hunk 4 above)
Comment (squirrelo) [inline qiita_db/metadata_template/util.py]: Done

**Comment by squirrelo** [comment]
All comments taken care of. Ready for another round

**Comment by antgonza** [comment]
:+1:

**Comment by squirrelo** [comment]
Hold on, checking something

**Comment by squirrelo** [comment]
Had to revert to ValueError because raising the QiitaDBError caused a 500 error on the interface side.

**Comment by josenavas** [comment]
Is the QiitaDBError not catched on the interface? If that's the case, I would suggest catching it rather than rely on ValueError (I know, more work, but core more robust...)

**Comment by squirrelo** [comment]
Done

**Comment by josenavas** [comment]
:+1: @wasade can you review this again?

**Comment by ElDeveloper** [comment]
Yargh, just noticed this PR is missing tests .... :sob:

**Comment by squirrelo** [comment]
It does: qiita_db/metadata_template/test/test_util.py

**Comment by ElDeveloper** [comment]
I entirely missed that. :eyes:

**Pull request merged**